crates/cranelift/src/compiler/component.rs

```diff
@@ -231,6 +225,223 @@ impl Compiler {
+
+        self.abi_preamble(&mut builder, &offsets, vmctx, abi);
+
+        // The arguments this shim passes along to the libcall are:
+        //
+        //   * the vmctx
+        //   * a constant value for this `ResourceRep` intrinsic
+        //   * the wasm argument to unwrap
+        let mut host_args = Vec::new();
+        host_args.push(vmctx);
+        host_args.push(
+            builder
+                .ins()
+                .iconst(ir::types::I32, i64::from(resource.resource.as_u32())),
+        );
+        host_args.push(args[2]);
+
+        // Currently this only support resources represented by `i32`
+        assert_eq!(ty.returns()[0], WasmType::I32);
+        let (host_sig, offset) = host::resource_rep32(self, &mut builder.func);
+
+        let host_fn = self.load_libcall(&mut builder, &offsets, vmctx, offset);
+        let call = builder.ins().call_indirect(host_sig, host_fn, &host_args);
+        let result = builder.func.dfg.inst_results(call)[0];
+        self.abi_store_results(&mut builder, ty, block0, &[result], abi);
+
+        builder.finalize();
+        Ok(Box::new(compiler.finish()?))
+    }
+
+    fn compile_resource_drop_for_abi(
+        &self,
+        component: &Component,
+        resource: &ResourceDrop,
+        types: &ComponentTypes,
+        abi: Abi,
+    ) -> Result<Box<dyn Any + Send>> {
+        let pointer_type = self.isa.pointer_type();
+        let ty = &types[resource.signature];
+        let isa = &*self.isa;
+        let offsets = VMComponentOffsets::new(isa.pointer_bytes(), component);
+        let mut compiler = self.function_compiler();
+        let func = self.func(ty, abi);
+        let (mut builder, block0) = compiler.builder(func);
+
+        let args = self.abi_load_params(&mut builder, ty, block0, abi);
+        let vmctx = args[0];
+
+        self.abi_preamble(&mut builder, &offsets, vmctx, abi);
+
+        // The arguments this shim passes along to the libcall are:
+        //
+        //   * the vmctx
+        //   * a constant value for this `ResourceDrop` intrinsic
+        //   * the wasm handle index to drop
+        let mut host_args = Vec::new();
+        host_args.push(vmctx);
+        host_args.push(
+            builder
+                .ins()
+                .iconst(ir::types::I32, i64::from(resource.resource.as_u32())),
+        );
+        host_args.push(args[2]);
+
+        let (host_sig, offset) = host::resource_drop(self, &mut builder.func);
+        let host_fn = self.load_libcall(&mut builder, &offsets, vmctx, offset);
+        let call = builder.ins().call_indirect(host_sig, host_fn, &host_args);
+        let should_run_destructor = builder.func.dfg.inst_results(call)[0];
+
+        // Synthesize the following:
+        //
+        //      ...
+        //      brif should_run_destructor, run_destructor_block, return_block
+        //
+        //    run_destructor_block:
+        //      rep = ushr.i64 rep, 1
+        //      rep = ireduce.i32 rep
```

Comment:
This is probably not a huge optimization, but these two instructions to compute the new `rep` value can be emitted into `call_func_ref_block` instead so that they're not executed if there is no destructor.